server/pulp/server/managers/repo/unit_association.py

```diff
@@ -139,7 +139,30 @@ def associate_all_by_ids(self, repo_id, unit_type_id, unit_id_list):
         return unique_count
 
     @staticmethod
-    def associate_from_repo(source_repo_id, dest_repo_id, criteria=None,
+    def _units_from_criteria(source_repo, criteria):
```

Comment:
That sounds like a fine idea for another day. Making this support a more general case would require adding support for limit, skip and sort at a minimum, which takes the scope well beyond my 2-line bugfix. Thanks for the suggestion!